drivers/src/lms.rs

```diff
@@ -0,0 +1,476 @@
+    hasher.update(lms_identifier)?;
+    hasher.update(q)?;
+    hasher.update(&D_MESG.to_be_bytes())?;
+    hasher.update(nonce)?;
+    hasher.update(message)?;
+    hasher.finalize()?;
+    Ok(HashValue::from(digest))
+}
+
+pub fn candidate_ots_signature<const N: usize, const P: usize>(
+    algo_type: &LmotsAlgorithmType,
+    lms_identifier: &LmsIdentifier,
+    q: &[u8; 4],
+    signature: &LmotsSignature<N, P>,
+    message_digest: &HashValue<N>,
+) -> CaliptraResult<HashValue<N>> {
+    if algo_type != &signature.ots_type {
+        // println!("These have different ots types");
+        raise_err!(InvalidLmotsAlgorithmType);
+    }
+    let params = get_lmots_parameters(algo_type)?;
+    if params.p as usize != P {
+        raise_err!(InvalidPValue);
+    }
+    if params.n > 34 {
+        raise_err!(InvalidHashWidth);
+    }
+    if params.n as usize != N {
+        raise_err!(InvalidHashWidth);
+    }
+    let mut z = [HashValue::<N>::default(); P];
+    let mut message_hash_with_checksum = [0u8; 34]; // 2 extra bytes for the checksum. needs to be N+2
+    message_hash_with_checksum[..N].copy_from_slice(&message_digest.0[..N]);
+
+    let checksum_q = checksum(algo_type, &message_hash_with_checksum)?;
+    let be_checksum = checksum_q.to_be_bytes();
+    message_hash_with_checksum[N] = be_checksum[0];
+    message_hash_with_checksum[N + 1] = be_checksum[1];
+
+    // In order to reduce the number of copies allocate a single block of memory
+    // and update only the portions that update between iterations
+    let mut hash_block = [0u8; 55];
+    hash_block[0..16].clone_from_slice(lms_identifier);
+    hash_block[16..20].clone_from_slice(q);
+    for i in 0..params.p {
+        let a = coefficient(&message_hash_with_checksum, i as usize, params.w as usize);
+        let mut tmp = signature.y[i as usize];
+        let t_upper: u16 = (1 << params.w) - 1; // subtract with overflow?
+        let upper = t_upper as u8;
+        hash_block[20..22].clone_from_slice(&i.to_be_bytes());
+        for j in a..upper {
+            let mut digest = Array4x8::default();
+            let sha = Sha256::default();
+            let mut hasher = sha.digest_init(&mut digest)?;
+            hash_block[22] = j;
+            hash_block[23..23 + N].clone_from_slice(&tmp.0);
+            hasher.update(&hash_block[0..23 + N])?;
+            hasher.finalize()?;
+            tmp = HashValue::<N>::from(digest);
+        }
+        z[i as usize] = tmp;
+    }
+    let mut digest = Array4x8::default();
+    let sha = Sha256::default();
+    let mut hasher = sha.digest_init(&mut digest)?;
+    hasher.update(lms_identifier)?;
+    hasher.update(q)?;
+    hasher.update(&D_PBLC.to_be_bytes())?;
+    for t in z {
+        hasher.update(&t.0)?;
+    }
+    hasher.finalize()?;
+    let result = HashValue::<N>::from(digest);
+    Ok(result)
+}
+
+pub fn verify_lms_signature<const N: usize, const P: usize>(
+    tree_height: u8,
```

Comment:
nit: mind creating a struct for this?
LmsPublicKey<N> or something like that?

Reply:
I likely will. I am waiting to polish this top level API until @vsonims has looked at it and we can refactor the API with how the ROM will read in the image.